File: golang/sorting/sorting.go
// Implementation of several sorting algorithms in Go.
package sorting

import (
	"sync"
)

// Sorts the given slice of integer using bubble sort algorithm.
func BubbleSort(sl []int) {
	sliceLength := len(sl)

	for i := 0; i < sliceLength-1; i++ {
		for j := 0; j < sliceLength-i-1; j++ {
			if sl[j] > sl[j+1] {
				sl[j], sl[j+1] = sl[j+1], sl[j]
			}
		}
	}
}

// Sorts the given slice of integer using selection algorithm.
func SelectionSort(sl []int) {
	sliceLength := len(sl)

	for i := 0; i < sliceLength; i++ {
		// Find minimum element in slice [i+1:sliceLength]
		minimumIdx := i
		for j := i + 1; j < sliceLength; j++ {
			if sl[j] < sl[minimumIdx] {
				minimumIdx = j
			}
		}

		// Swap these elements
		sl[i], sl[minimumIdx] = sl[minimumIdx], sl[i]
	}
}

// Sorts the given slice of integer using insertion algorithm.
func InsertionSort(sl []int) {
	sliceLength := len(sl)

	for i := 1; i < sliceLength; i++ {
		for j := i - 1; j >= 0; j-- {
			if sl[j] > sl[j+1] {
				sl[j], sl[j+1] = sl[j+1], sl[j]
			} else {
				break
			}
		}
	}
}

// Helper for function for merge sort
func merge(firstHalf []int, secondHalf []int) []int {
	firstHalfLength := len(firstHalf)
	secondHalfLength := len(secondHalf)
	result := make([]int, 0, firstHalfLength+secondHalfLength)

	firstHalfIdx, secondHalfIdx := 0, 0

	// Iterate through both and put smaller value
	for firstHalfIdx < firstHalfLength && secondHalfIdx < secondHalfLength {
		if firstHalf[firstHalfIdx] < secondHalf[secondHalfIdx] {
			result = append(result, firstHalf[firstHalfIdx])
			firstHalfIdx++
		} else {
			result = append(result, secondHalf[secondHalfIdx])
			secondHalfIdx++
		}
	}

	// Pick from first, if left
	for firstHalfIdx < firstHalfLength {
		result = append(result, firstHalf[firstHalfIdx])
		firstHalfIdx++
	}

	// Pick from second, if left
	for secondHalfIdx < secondHalfLength {
		result = append(result, secondHalf[secondHalfIdx])
		secondHalfIdx++
	}

	return result
}

// Sorts the given slice of integer using merge algorithm.
func MergeSort(sl []int) {
	sliceLength := len(sl)
	if sliceLength == 1 {
		return
	}

	var middle int = (sliceLength - 1) / 2

	firstHalf := sl[0 : middle+1]
	secondHalf := sl[middle+1 : sliceLength]

	MergeSort(firstHalf)
	MergeSort(secondHalf)

	copy(sl, merge(firstHalf, secondHalf))
}

// Sorts the given slice of integer using simple parallel merge algorithm.
func MergeSortParallel(sl []int) {
	wg := &sync.WaitGroup{}
	sliceLength := len(sl)
	if sliceLength == 1 {
		return
	}

	var middle int = (sliceLength - 1) / 2

	firstHalf := sl[0 : middle+1]
	secondHalf := sl[middle+1 : sliceLength]

	wg.Add(1)
	go func() {
		defer wg.Done()
		MergeSort(firstHalf)
	}()
	wg.Add(1)
	go func() {
		defer wg.Done()
		MergeSort(secondHalf)
	}()

	wg.Wait()
	copy(sl, merge(firstHalf, secondHalf))
}

// Helper function for sorting according to pivot element for quicksort.
// Returns the final position of pivot element
func doPivot(sl []int, start, end int) int {
	pivotIdx := end - 1
	swapIdx := start

	for i := start; i < end-1; i++ {
		if sl[pivotIdx] > sl[i] {
			sl[i], sl[swapIdx] = sl[swapIdx], sl[i]
			swapIdx++
		}
	}

	sl[pivotIdx], sl[swapIdx] = sl[swapIdx], sl[pivotIdx]
	return swapIdx
}

// Helper recursive function for quicksort
func recursiveQuickSort(sl []int, start, end int) {
	if start < end-1 {
		pivotIdx := doPivot(sl, start, end)
		recursiveQuickSort(sl, start, pivotIdx)
		recursiveQuickSort(sl, pivotIdx, end)
	}
}

// Sorts the given slice of integer using quicksort algorithm. Last element is chosen
// as pivot.
func QuickSort(sl []int) {
	recursiveQuickSort(sl, 0, len(sl))
}

File: c/src/sorting.c
#include <stdlib.h>

#include "sorting.h"
#include "string.h"

// Simple macro for swapping values
#define swap(type, rhs, lhs) do {       \
    type tmp = rhs;                     \
    rhs = lhs;                          \
    lhs = tmp;                          \
} while (0)                             \

void bubbleSort(ElemT* arr, int size) {
    for (int i = 0; i < size; i++) {
        for (int j = i + 1; j < size; j++) {
            if (arr[i] > arr[j])
                swap(ElemT, arr[i], arr[j]);
        }
    }
}

void selectionSort(ElemT* arr, int size) {
    int minIdx = 0;
    for (int i = 0; i < size; ++i) {
        minIdx = i;
        for (int j = i + 1; j < size; ++j) {
            if (arr[j] < arr[minIdx])
                minIdx = j;
        }

        swap(ElemT, arr[i], arr[minIdx]);
    }
}

void insertionSort(ElemT* arr, int size) {
    for (int i = 1; i < size; ++i) {
        for (int j = i - 1; j >= 0; --j) {
            if (arr[j] > arr[j + 1])
                swap(ElemT, arr[j], arr[j + 1]);
            else
                break;
        }
    }
}

// Merge two parts into one array, size of the result is the sum of given two
static ElemT* merge(ElemT* lhs, int lsize, ElemT* rhs, int rsize) {
    ElemT* result = (ElemT*) calloc(lsize + rsize, sizeof(ElemT));
    int resultIdx = 0;
    if (result == NULL) {
        printf("merge(): calloc error\n");
        return NULL;
    }

    // Iterate both parts
    int ldx = 0, rdx = 0;
    while (ldx < lsize && rdx < rsize) {
        if (lhs[ldx] > rhs[rdx]) {
            result[resultIdx++] = rhs[rdx++];
        } else {
            result[resultIdx++] = lhs[ldx++];
        }
    }

    // Collect from the rest
    while (ldx < lsize) {
        result[resultIdx++] = lhs[ldx++];
    }

    while (rdx < rsize) {
        result[resultIdx++] = rhs[rdx++];
    }

    return result;
}

// [start;end) 
static void recursiveMergeSort(ElemT* arr, int start, int end) {
    if (end - start <= 0) {
        printf("recursiveMergeSort(): end - start <= 0\n");
        return ;
    }

    if (end - start == 1)
        return ;

    int middle = (end - start - 1) / 2 + 1;

    // Sort recursively two parts
    recursiveMergeSort(arr, start, start + middle);
    recursiveMergeSort(arr, start + middle, end);

    // Merge results
    ElemT* sorted = merge(arr + start, middle, arr + start + middle, end - start - middle);
    // Copy result to the input array
    memcpy(arr + start, sorted, (end - start) * sizeof(ElemT));
}

void mergeSort(ElemT* arr, int size) {
    recursiveMergeSort(arr, 0, size);
}

int doPivot(ElemT* arr, int start, int end) {
    int pivotIdx = end - 1;
    int swapIdx = start;

    for (int i = start; i < end - 1; ++i)
        if (arr[i] < arr[pivotIdx]) {
            swap(ElemT, arr[i], arr[swapIdx]);
            ++swapIdx;
        }

    swap(ElemT, arr[swapIdx], arr[pivotIdx]);
    return swapIdx;
}

void recursiveQuickSort(ElemT* arr, int start, int end) {
    if (start < end - 1) {
        int pivotIdx = doPivot(arr, start, end);
        recursiveQuickSort(arr, start, pivotIdx);
        recursiveQuickSort(arr, pivotIdx, end);
    }
}

void quickSort(ElemT* arr, int size) {
    recursiveQuickSort(arr, 0, size);
}